web/server/www/scripts/codecheckerviewer/filter/ManageFilters.js

```diff
@@ -0,0 +1,290 @@
+          if (filterValues[key] === null) {
+            delete filterValues[key];
+          }
+        });
+
+        CC_SERVICE.addReportFilter(filterName, JSON.stringify(filterValues),
+        function (res) {
+          if (res) {
+            that.hide();
+          }
+        }).fail(function (xhr) { util.handleAjaxFailure(xhr); });
+      }
+    },
+
+    postCreate : function () {
+      this.inherited(arguments);
+
+      var that = this;
+
+      var contentArea = dom.create('div', {
+        class: 'dijitDialogPaneContentArea'
+      }, this.containerNode);
+
+      this.filterName = new ValidationTextBox({
+        class       : 'form-input',
+        name        : 'filterName',
+        placeholder : 'Filter name...',
+        required    : true
+      }).placeAt(contentArea);
+
+      var actionBar = dom.create('div', {
+        class: 'dijitDialogPaneActionBar'
+      }, this.containerNode);
+
+      new Button({
+        label: 'Save',
+        onClick: function() {
+          that.save();
+        }
+      }).placeAt(actionBar);
+
+      new Button({
+        label: 'Cancel',
+        onClick: function() {
+          that.hide();
+        }
+      }).placeAt(actionBar);
+    },
+
+    onShow : function () {
+      this.filterName.reset();
+    }
+  });
+
+  function removeBtnFormatter(listOfFilters) {
+    return function (filterId) {
+      return new Button({
+        label: 'Remove',
+        iconClass : 'customIcon delete',
+        onClick : function () {
+          CC_SERVICE.removeReportFilter(filterId, function (res) {
+            if (res) {
+              listOfFilters.populateFilters();
+            }
+          }).fail(function (xhr) { util.handleAjaxFailure(xhr); });
+        }
+      });
+    };
+  }
+
+  var ListOfFilters = declare(DataGrid, {
+    constructor : function () {
+      this.store = new ItemFileWriteStore({
+        data : { identifier : 'id', items : [] }
+      });
+
+      this.structure = [
+        { name : 'Name', field : 'name', cellClasses : 'link', width : '100%'},
+        { name : '&nbsp;', field : 'remove', width : '100px', formatter : removeBtnFormatter(this) }
+      ];
+
+      this.focused = true;
+      this.selectable = true;
+      this.keepSelection = true;
+      this.escapeHTMLInData = false;
+      this.autoHeight = true;
+    },
+
+    addFilter : function (filter) {
+      this.store.newItem({
+        id : filter.id,
+        name : filter.name,
+        value : filter.value,
+        remove : filter.id
+      });
+    },
+
+    populateFilters : function () {
+      var that = this;
+
+      // Remove the old items.
+      this.store.fetch({
+        onComplete : function (filters) {
+          filters.forEach(function (filter) {
+            that.store.deleteItem(filter);
+          });
+          that.store.save();
+        }
+      });
+
+      // Load and render the items.
+      CC_SERVICE.getReportFilters(null, function (filters) {
```

Comment:
I saved a filter to my list. When I click on "Load" button, the saved filter is shown properly. When I close the window and click "Load" again, the filter is shown twice. On the 3rd "Load" an exception is shown. If I put this `getReportFilters()` function call to the magic `setTimeout()` then it works.

The `fetch()` and `save()` functions are async, but even if I place this `getReportFilters()` to their `onComplete()`, it doesn't help.

Reply:
i did not experience similar problem.